specs/governance/alligator.md

```diff
@@ -0,0 +1,61 @@
+# Alligator
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Overview](#overview)
+- [Subdelegations](#subdelegations)
+- [Token Transfer Hook](#token-transfer-hook)
+- [View functions](#view-functions)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+The `Alligator` contract implements subdelegations that can be used by tokens contracts that inherit
+[`GovernanceToken`](gov-token.md). Subdelegations allow for advanced delegation use cases, such as partial,
+time-constrained & block-based delegations, relative & fixed allowances, and custom rules. The `Alligator` contract
+requires migrating the delegation state from the token contract to itself through a hook-based approach. Concretely,
+token contracts call the `Alligator` contract's `afterTokenTransfer` function after a token transfer. This allows
+the `Alligator` contract to consume the hooks and update its delegation and checkpoint mappings accordingly. If either
+of the addresses passed as arguments to `afterTokenTransfer`, `from` and `to`, have not been migrated to the `Alligator`
+contract, `Alligator` copies the address's delegation and checkpoint data from the token contract to its own state.
+
+## Subdelegations
+
+To support subdelegations for multiple token contracts, the `Alligator` contract MUST implement storage references to
+store delegations and checkpoints for addresses. Specifically, the `Alligator` contract MUST implement the equivalent
+of the following mappings from `GovernanceToken` contract: `_delegates`, `_checkpoints`, `_totalSupplyCheckpoints`.
+
+The contract MUST also provide methods for subdelegation voting power to another address. Concretely, the `Alligator`
+contract MUST implement a `subdelegate` function which takes a `to` address, and a subdelegation rule. A subdelegation
+rule MUST contain the following fields:
+
+- `maxRedelegations`: The maximum number of times the delegated votes can be redelegated.
+- `blocksBeforeVoteCloses`: The number of blocks before the vote closes that the delegation is valid.
+- `notValidBefore`: The timestamp after which the delegation is valid.
+- `notValidAfter`: The timestamp before which the delegation is valid.
+- `allowanceType` The type of allowance. If `Absolute`, the amount of votes delegated is fixed. If `Relative`, the
```

Comment:
Are the only 2 types absolute and relative?

Reply:
existing contracts only define absolute & relative. do we need anything else? If not, I'm not sure it would be worth it to add more allowance types under the scope of this PR